[Tianrun] iP

trgao opened the pull request:
# ~~Duke~~ Cortana
> "Your mind is for having ideas, not holding them. " - David Allen ([source](https://dansilvestre.com/productivity-quotes))
## Cortana frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- SUPER FAST to use
## All you need to do is,
1. Download it from [here](https://github.com/trgao/ip/releases/tag/A-Jar)
2. Double click it
3. Add your tasks
4. Let it manage your tasks for you 💯 
And its **free**

## Features:
- [x] Managing todo, event and deadline tasks
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

trgao committed:
Change name of chatbot

trgao committed:
Refactor code

trgao committed:
Add echo functionality

trgao committed:
Refactor code

trgao committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested

trgao committed:
Refactor code into functions

trgao committed:
Refactor tasks into a new class

trgao committed:
Change snakecase to camelcase

trgao committed:
Change snakecase to camelcase

trgao committed:
Add the ability to mark tasks as done and the ability to change the status back to not done

trgao committed:
Add support for tracking three types of tasks: todo, deadline and event

trgao committed:
Add exception handling

trgao committed:
Add delete functionality

trgao committed:
Add automated text ui testing

trgao committed:
Fix switch case style

trgao committed:
Refactor UI methods to new class

trgao committed:
Refactor Parser methods to new class

trgao committed:
Refactor TaskList into a new class

trgao committed:
Add save functionality

trgao committed:
Revert "Add save functionality"

This reverts commit 30144f3519c2d29e3a59ae479248528c56a5d36a.

trgao committed:
Create duke.txt

trgao committed:
Revert "Create duke.txt"

This reverts commit 7252706b473c0110be950d74170321b6d4d7d4e3.

trgao committed:
Add save functionality

trgao committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save functionality

trgao committed:
Update gitignore file

trgao committed:
Store date as LocalDate in task objects

trgao committed:
Remove unused import statement

trgao committed:
Add get deadlines and events by date functionality

trgao committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add get deadlines and events by date functionality
  Remove unused import statement
  Store date as LocalDate in task objects

trgao committed:
Refactor files into duke package

trgao committed:
Refactor tasks into duke.task package

trgao committed:
Merge remote-tracking branch 'origin/add-gradle-support'

* origin/add-gradle-support:
  Bump gradle and lib version
  Add Gradle support

trgao committed:
Add JUnit tests

trgao committed:
Build jar file

trgao committed:
Add header comments to all non-private classes/methods, and non-trivial private methods

trgao committed:
Merge branch 'A-JavaDoc'

* A-JavaDoc:
  Add header comments to all non-private classes/methods, and non-trivial private methods

trgao committed:
Change error message

trgao committed:
Add find by description functionality

trgao committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find by description functionality

trgao committed:
Fix style problems

trgao committed:
Add checkstyle

trgao committed:
Fix style issues

trgao committed:
Merge branch 'A-CheckStyle'

* A-CheckStyle:
  Fix style issues
  Add checkstyle

trgao committed:
Add skeleton javafx window

trgao committed:
Add basic GUI

maypfv commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,91 @@
+        }
+        return taskList;
+    }
+}
```
I like the readability of the methods here. Good! 👍

maypfv commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,331 @@
+        } else {
+            return taskListToString(taskList);
+        }
+    }
```
I like the use of streams to find the task description.

trgao committed:
Fix style issues

trgao committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Fix style issues
  Add basic GUI
  Add skeleton javafx window

maypfv commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,331 @@
+        if (index < 0 || index >= size) {
+            throw new DukeException(String.format("!!!: Please provide a valid number, "
+                    + "there are currently %d items in the list", size));
+        }
```
Perhaps good to handle such exceptions in parser instead of TaskList for better readability on how to handle tasks in tasklist.

maypfv commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,331 @@
+        }
+        String dueDate = line.substring(line.indexOf("/by") + 4);
+        if (dueDate.isBlank()) {
+            throw new DukeException("!!!: Please provide a due date");
```
As mentioned, perhaps you can handle such exceptions in Parser.

trgao committed:
Add save to file upon application close

maypfv commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,331 @@
+        }
+        return output.toString();
+    }
+}
```
Perhaps using hashtable could be premature optimization.

maypfv reviewed (commented):
Overall looks really good! Just minor suggestions for readability.

trgao committed:
Refactor GUI into controllers and views

trgao committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Refactor GUI into controllers and views

trgao committed:
Fix merge conflicts

trgao committed:
Fix build.gradle file

YeoBohShin commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,125 @@
+/**
+ * Represents the object that displays output from the chatbot.
+ */
+public class UI {
```
i like the naming of the methods, easy to understand

YeoBohShin commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,35 @@
+     * @param dueDate The due date of the task
+     * @throws DateTimeParseException if dueDate is not in the correct format
+     */
+    public Deadline(String description, String dueDate) throws DateTimeParseException {
```
Maybe can consider parsing into a local date before accepting as a parameter?

YeoBohShin commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,331 @@
+        if (description.isBlank()) {
+            throw new DukeException("!!!: Please provide a description to search for");
+        }
+        ArrayList<Task> taskList = list.stream()
+                .filter(task -> task.description.contains(description))
+                .collect(Collectors.toCollection(ArrayList::new));
```
Nice use of streams here!

YeoBohShin reviewed (commented):
Overall good job on the ip. Keep it up!

trgao committed:
Add assert line

trgao committed:
Merge pull request #1 from trgao/A-Assertions

Add assert line

trgao committed:
Add github actions workflow

trgao committed:
Update github workflow actions file

trgao committed:
Change find date implementation

There are multiple hashtables keeping track of different types of tasks with different sets of dates.

This is unnecessariliy complicated, as trying to find different types of tasks on different dates requires multiple commands.

Let's combine all of the tables into one single hashtable to keep track of the dates.

This is so that a date can query for all different types of tasks using just one command, which will make the chatbot easier to use.

trgao committed:
Add find free dates functionality

There is no way for the user to find what are the next available dates in his/her schedule.

If the user already has an event on a certain date, he/she might not want to add another event on the same date.

Let's add a functionality for the user to find what are the next available dates with the command "freedates", which will return the 5 closest date to the current date that is free.

trgao committed:
Fix test cases for gradle

trgao committed:
Merge pull request #2 from trgao/B-FindFreeTimes

Add find free dates functionality

trgao committed:
Add files via upload

trgao committed:
Update README.md

trgao committed:
Update README.md

trgao committed:
Delete docs/UI.png

trgao committed:
Add files via upload

trgao committed:
Add dependencies for JavaFX

trgao committed:
Merge branch 'master' of https://github.com/trgao/ip

* 'master' of https://github.com/trgao/ip:
  Add files via upload
  Delete docs/UI.png

trgao committed:
Update build.gradle file

trgao committed:
Update README.md

trgao committed:
Update README.md

trgao committed:
Update README.md

trgao committed:
Update README.md

trgao committed:
Update README.md

trgao committed:
Add files via upload

trgao committed:
Update _config.yml

trgao committed:
Update _config.yml

trgao committed:
Update _config.yml

trgao committed:
Update _config.yml

trgao committed:
Update README.md
